File: 04-practice/ex05-shapes/solution.js
class Circle {
  constructor(radius) {
    this.radius = radius;
  }
  area() {
    return Math.PI * this.radius ** 2;
  }
}

class Rectangle {
  constructor(width, height) {
    this.width = width;
    this.height = height;
  }
  area() {
    return this.width * this.height;
  }
}

function largest(shapes) {
  if (shapes.length === 0) return null;
  return shapes.reduce((best, s) => (s.area() > best.area() ? s : best));
}

function assert(cond, label) {
  console.log(cond ? `OK   ${label}` : `FAIL ${label}`);
}

const c = new Circle(1);
const r = new Rectangle(3, 4);
assert(Math.abs(c.area() - Math.PI) < 1e-9, "circle area");
assert(r.area() === 12, "rect area");
assert(largest([c, r]) === r, "largest");
assert(largest([]) === null, "empty");

File: 04-practice/ex05-shapes/solution.py
from math import pi


class Circle:
    def __init__(self, radius):
        self.radius = radius

    def area(self):
        return pi * self.radius ** 2


class Rectangle:
    def __init__(self, width, height):
        self.width = width
        self.height = height

    def area(self):
        return self.width * self.height


def largest(shapes):
    if not shapes:
        return None
    return max(shapes, key=lambda s: s.area())


def assert_(cond, label):
    print(f"OK   {label}" if cond else f"FAIL {label}")


c = Circle(1)
r = Rectangle(3, 4)
assert_(abs(c.area() - pi) < 1e-9, "circle area")
assert_(r.area() == 12, "rect area")
assert_(largest([c, r]) is r, "largest")
assert_(largest([]) is None, "empty")
